### PR TITLE
Add device inclusive scan with init_value

### PR DESCRIPTION
**Pull request by gonidelis**
Still working towards but not completely solving #698.

Builds on top of just merged #1749 to add `cub::DeviceScan::InclusiveScan` overload that accepts initial value without braking the existing API (non init-value based).

Unit and API tests included.

Needed to build on top of it for the `thrust::inclusive_scan` overloads.

### CODE REVIEW

**Review by gevtushenko — CHANGES_REQUESTED** [review]
Fitting a new overload in CUB without breaking something is a hard task. We'll have to figure out an alternative approach and probably extend our tests to cover these corner cases before merging this PR.

File: cub/test/catch2_test_device_scan.cu [hunk 1]
```diff
@@ -210,6 +210,47 @@ CUB_TEST("Device scan works with all device interfaces", "[scan][device]", full_
     }
   }
 
+// To call DeviceScan::InclusiveScan with initial value, the cudaStream_t argument needs
+// to be passed explicitly in order to maintain backwards compatibility. Since
+// DECLARE_LAUNCH_WRAPPER passes stream by default as the last argument for TEST_LAUNCH = 2
+// we need to skip these set of tests.
+#if TEST_LAUNCH != 2
```
Comment (gevtushenko) [inline cub/test/catch2_test_device_scan.cu:217]: important: we can properly test this with something along the following lines:
```cpp
template <typename InputIteratorT, typename OutputIteratorT, typename ScanOpT, typename InitT>
CUB_RUNTIME_FUNCTION static cudaError_t inclusive_scan_init_wrapper(
  void* d_temp_storage,
  size_t& temp_storage_bytes,
  InputIteratorT d_in,
  OutputIteratorT d_out,
  ScanOpT scan_op,
  int num_items,
  InitT init,
  cudaStream_t stream = 0)
{
  return cub::DeviceScan::InclusiveScan(
    d_temp_storage, temp_storage_bytes, d_in, d_out, scan_op, num_items, stream, init);
}
DECLARE_LAUNCH_WRAPPER(inclusive_scan_init_wrapper, device_inclusive_scan_with_init);
```

```suggestion
```

File: cub/test/catch2_test_device_scan.cu [hunk 2]
```diff
@@ -210,6 +210,47 @@ CUB_TEST("Device scan works with all device interfaces", "[scan][device]", full_
...
+    c2h::host_vector<input_t> host_items(in_items);
+    c2h::host_vector<output_t> expected_result(num_items);
+    compute_inclusive_scan_reference(
+      host_items.cbegin(), host_items.cend(), expected_result.begin(), scan_op, accum_t{});
+
+    // Run test
+    c2h::device_vector<output_t> out_result(num_items);
+    auto d_out_it = thrust::raw_pointer_cast(out_result.data());
+    using init_t  = cub::detail::value_t<decltype(unwrap_it(d_out_it))>;
+    cudaStream_t stream{};
```
Comment (gevtushenko) [inline cub/test/catch2_test_device_scan.cu:236]: suggestion: with the change above, we don't need the stream
```suggestion
```

File: cub/test/catch2_test_device_scan.cu [hunk 3]
```diff
@@ -210,6 +210,47 @@ CUB_TEST("Device scan works with all device interfaces", "[scan][device]", full_
...
+
+    device_inclusive_scan(unwrap_it(d_in_it), unwrap_it(d_out_it), scan_op, num_items, stream, init_t{});
+
```
Comment (gevtushenko) [inline cub/test/catch2_test_device_scan.cu:239]: critical: the binary operator is sum, which makes this initial value an identity element. If this overload doesn't do with the initial value, we'll never see that. Let's follow what other tests do with respect to initial value.

File: cub/test/catch2_test_device_scan.cu [hunk 4]
```diff
@@ -210,6 +210,47 @@ CUB_TEST("Device scan works with all device interfaces", "[scan][device]", full_
...
+    using accum_t = cub::detail::accumulator_t<op_t, input_t, input_t>;
+
+    // Scan operator
+    auto scan_op = unwrap_op(reference_extended_fp(d_in_it), op_t{});
+
+    // Prepare verification data
+    c2h::host_vector<input_t> host_items(in_items);
+    c2h::host_vector<output_t> expected_result(num_items);
+    compute_inclusive_scan_reference(
+      host_items.cbegin(), host_items.cend(), expected_result.begin(), scan_op, accum_t{});
```
Comment (gevtushenko) [inline cub/test/catch2_test_device_scan.cu:230]: important: use same initial value in reference and CUB invocations.

File: cub/cub/device/device_scan.cuh [hunk 5]
```diff
@@ -1164,6 +1164,110 @@ struct DeviceScan
       d_temp_storage, temp_storage_bytes, d_in, d_out, scan_op, NullType(), num_items, stream);
   }
 
+  //! @rst
+  //! Computes a device-wide inclusive prefix scan using the specified binary ``scan_op`` functor.
+  //! The result of applying the ``scan_op`` binary operator to ``init_value`` value and ``*d_in``
+  //! is assigned to ``*d_dout``.
```
Comment (gevtushenko) [inline cub/cub/device/device_scan.cuh:1170]:
```suggestion
  //! is assigned to ``*d_out``.
```

File: cub/test/catch2_test_device_scan.cu [hunk 6]
```diff
@@ -210,6 +227,40 @@ CUB_TEST("Device scan works with all device interfaces", "[scan][device]", full_
...
+    REQUIRE(expected_result == out_result);
+
+    // Run test in-place
+    _CCCL_IF_CONSTEXPR (std::is_same<input_t, output_t>::value)
+    {
+      device_inclusive_scan_with_init(unwrap_it(d_in_it), unwrap_it(d_in_it), scan_op, num_items, init_t{});
+
+      // Verify result
+      REQUIRE(expected_result == in_items);
+    }
```
Comment (gevtushenko) [inline cub/test/catch2_test_device_scan.cu:261]: suggestion: many types that this test is going to be invoked with are unsigned, so having op=min + init=0 = fill(out, 0). I'd preffer to use sum, so that we incorporated initial value into result rather than mask it:
```suggestion
    using op_t    = cub::Sum;
    using accum_t = cub::detail::accumulator_t<op_t, input_t, input_t>;

    // Scan operator
    auto scan_op = unwrap_op(reference_extended_fp(d_in_it), op_t{});

    // Prepare verification data
    c2h::host_vector<input_t> host_items(in_items);
    c2h::host_vector<output_t> expected_result(num_items);

    // Run test
    c2h::device_vector<output_t> out_result(num_items);
    auto d_out_it = thrust::raw_pointer_cast(out_result.data());
    accum_t init_value{};
    init_default_constant(init_value);
    compute_inclusive_scan_reference(
      host_items.cbegin(), host_items.cend(), expected_result.begin(), scan_op, init_value);

    device_inclusive_scan_with_init(unwrap_it(d_in_it), unwrap_it(d_out_it), scan_op, num_items, init_value);

    // Verify result
    REQUIRE(expected_result == out_result);

    // Run test in-place
    _CCCL_IF_CONSTEXPR (std::is_same<input_t, output_t>::value)
    {
      device_inclusive_scan_with_init(unwrap_it(d_in_it), unwrap_it(d_in_it), scan_op, num_items, init_value);

      // Verify result
      REQUIRE(expected_result == in_items);
    }
```

File: cub/cub/device/device_scan.cuh [hunk 7]
```diff
@@ -1164,6 +1164,110 @@ struct DeviceScan
...
+  template <typename InputIteratorT, typename OutputIteratorT, typename ScanOpT, typename InitValueT>
+  CUB_RUNTIME_FUNCTION static cudaError_t InclusiveScan(
+    void* d_temp_storage,
+    size_t& temp_storage_bytes,
+    InputIteratorT d_in,
+    OutputIteratorT d_out,
+    ScanOpT scan_op,
+    int num_items,
+    cudaStream_t stream,
+    InitValueT init_value)
```
Comment (gevtushenko) [inline cub/cub/device/device_scan.cuh:1244]: critical: we have a debug-synchronous overload expecting boolean parameter after stream. If user passed 0 instead of bool, we'd select a different overload, potentially changing a behavior of the algorithm: 

```cpp
// 0 used to mean debug synchronous but now it's initial value
cub::DeviceScan::InclusiveScan(d_temp_storage, temp_storage_bytes, d_data, d_data, max{}, num_items, stream, 0 /* was debug_synchronous but now init_value*/);
```

Opposite problem happens when initial value is bool. In this case we invoke deprecated API instead of the new API:

```cpp
bool init_value = 0;
cub::DeviceScan::InclusiveScan(d_temp_storage, temp_storage_bytes, d_data, d_data, cub::Sum{}, num_items, stream, init_value /* but treated as debug_synchronous */);
```

We should find another way

**Review by gevtushenko — CHANGES_REQUESTED** [review]

File: cub/cub/device/device_scan.cuh [hunk 8]
```diff
@@ -1164,6 +1164,110 @@ struct DeviceScan
...
+  {
+    CUB_DETAIL_NVTX_RANGE_SCOPE_IF(d_temp_storage, "cub::DeviceScan::InclusiveScan");
```
Comment (gevtushenko) [inline cub/cub/device/device_scan.cuh:1246]: suggestion:
```suggestion
    CUB_DETAIL_NVTX_RANGE_SCOPE_IF(d_temp_storage, "cub::DeviceScan::InclusiveScanInit");
```

File: cub/cub/device/device_scan.cuh [hunk 9]
```diff
@@ -1164,6 +1164,110 @@ struct DeviceScan
...
+  template <typename InputIteratorT, typename OutputIteratorT, typename ScanOpT, typename InitValueT>
+  CUB_RUNTIME_FUNCTION static cudaError_t InclusiveScanInit(
+    void* d_temp_storage,
+    size_t& temp_storage_bytes,
+    InputIteratorT d_in,
+    OutputIteratorT d_out,
+    ScanOpT scan_op,
+    int num_items,
+    cudaStream_t stream,
+    InitValueT init_value)
```
Comment (gevtushenko) [inline cub/cub/device/device_scan.cuh:1244]: important: there's no reason to break consistency given the fact that we went with different function name. Let's have stream as last parameter and default it to zero.

**Review by gevtushenko — CHANGES_REQUESTED** [review]
Last minor change and let's merge it

File: cub/test/catch2_test_device_scan_api.cu [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+#include <thrust/device_vector.h>
+#include <thrust/host_vector.h>
+
+#include "catch2_test_helper.h"
+
+CUB_TEST("Device inclusive scan works", "[scan][device]")
+{
+  // Determine temporary device storage requirements for inclusive scan
+  cudaStream_t stream{};
+  REQUIRE(cudaSuccess == cudaStreamCreate(&stream));
```
Comment (bernhardmgruber) [inline cub/test/catch2_test_device_scan_api.cu:39]: Question: Why do we need this stream?

File: cub/test/catch2_test_device_scan_api.cu [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
```
Comment (bernhardmgruber) [inline cub/test/catch2_test_device_scan_api.cu:26]: Important: IIRC we discussed recently again that all new code should be licensed under Apache 2 with LLVM exception. In the team VC on Monday I proposed this license header:
```
// SPDX-FileCopyrightText: Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```
and the present team members agreed. But let's bring this up in the team meeting today

File: cub/test/catch2_test_device_scan.cu [hunk 12]
```diff
@@ -33,12 +33,29 @@
...
+  InputIteratorT d_in,
+  OutputIteratorT d_out,
+  ScanOpT scan_op,
+  InitT init,
+  int num_items,
+  cudaStream_t stream = 0)
+{
+  return cub::DeviceScan::InclusiveScanInit(
+    d_temp_storage, temp_storage_bytes, d_in, d_out, scan_op, init, num_items, stream);
+}
```
Comment (bernhardmgruber) [inline cub/test/catch2_test_device_scan.cu:56]: Question: I am confused, because `InclusiveScanInit` takes `stream` as the last argument, as it should. Why is this necessary?

File: cub/test/catch2_test_device_scan_api.cu (hunk 11 above)
Comment (bernhardmgruber) [reply cub/test/catch2_test_device_scan_api.cu:26]: I just discussed this with @jrhemstad and he suggested to stay with what other CUB files are doing for now. So you can keep the license header as is.

File: cub/test/catch2_test_device_scan_api.cu (hunk 10 above)
Comment (gonidelis) [reply cub/test/catch2_test_device_scan_api.cu:39]: amazing catch

File: cub/test/catch2_test_device_scan.cu (hunk 12 above)
Comment (gonidelis) [reply cub/test/catch2_test_device_scan.cu]: I completely deleted this wrapper, you were right, thanks a lot.

File: cub/cub/device/device_scan.cuh [hunk 13]
```diff
@@ -1164,6 +1164,110 @@ struct DeviceScan
...
+  //! - Supports non-commutative scan operators.
+  //! - Results are not deterministic for pseudo-associative operators (e.g.,
+  //!   addition of floating-point types). Results for pseudo-associative
+  //!   operators may vary from run to run. Additional details can be found in
+  //!   the @lookback description.
+  //! - When ``d_in`` and ``d_out`` are equal, the scan is performed in-place. The
+  //!   range ``[d_in, d_in + num_items)`` and ``[d_out, d_out + num_items)``
+  //!   shall not overlap in any other way.
+  //! - Note that because initial value is the last argument the cudaStream_t
+  //!   argument also needs be passed explicitly.
```
Comment (bernhardmgruber) [inline cub/cub/device/device_scan.cuh:1181]: Suggestion: missing article and wrong docmentation. The stream is the last argument in your function, not the initial value.
```suggestion
  //! - Note that because the initial value is the last argument the cudaStream_t
  //!   argument also needs be passed explicitly.
```

File: cub/cub/device/dispatch/dispatch_scan.cuh [hunk 14]
```diff
@@ -234,7 +239,8 @@ template <typename InputIteratorT,
                                                                    cub::detail::value_t<InputIteratorT>,
                                                                    typename InitValueT::value_type>,
                                                   cub::detail::value_t<InputIteratorT>>,
-          typename SelectedPolicy = DeviceScanPolicy<AccumT, ScanOpT>>
+          typename SelectedPolicy = DeviceScanPolicy<AccumT, ScanOpT>,
+          bool ForceInclusive     = false>
```
Comment (bernhardmgruber) [inline cub/cub/device/dispatch/dispatch_scan.cuh:243]: Question: we usually put the policy as the last parameter. Why is it necessary to break the design here?

File: cub/cub/device/device_scan.cuh [hunk 15]
```diff
@@ -1164,6 +1164,110 @@ struct DeviceScan
...
+  //!
+  //! @tparam OutputIteratorT
+  //!   **[inferred]** Random-access output iterator type for writing scan outputs @iterator
+  //!
+  //! @tparam ScanOpT
+  //!   **[inferred]** Binary scan functor type having member `T operator()(const T &a, const T &b)`
+  //!
+  //! @tparam InitValueT
+  //!  **[inferred]** Type of the `init_value` used Binary scan functor type
+  //!   having member `T operator()(const T &a, const T &b)`
```
Comment (bernhardmgruber) [inline cub/cub/device/device_scan.cuh:1208]: Suggestion: Improve wording. I cannot follow this description is trying to tell me.

File: cub/cub/device/device_scan.cuh [hunk 16]
```diff
@@ -1164,6 +1164,110 @@ struct DeviceScan
...
+  //!
+  //! @param[in] d_temp_storage
+  //!   Device-accessible allocation of temporary storage.
+  //!   When `nullptr`, the required allocation size is written to
+  //!   `temp_storage_bytes` and no work is done.
+  //!
+  //! @param[in,out] temp_storage_bytes
+  //!   Reference to size in bytes of `d_temp_storage` allocation
```
Comment (bernhardmgruber) [inline cub/cub/device/device_scan.cuh:1216]: Suggestion
```suggestion
  //!   Reference to the size in bytes of the `d_temp_storage` allocation
```

File: cub/cub/device/device_scan.cuh [hunk 17]
```diff
@@ -1164,6 +1164,110 @@ struct DeviceScan
...
+  //!   Random-access iterator to the input sequence of data items
+  //!
+  //! @param[out] d_out
+  //!   Random-access iterator to the output sequence of data items
+  //!
+  //! @param[in] scan_op
+  //!   Binary scan functor
+  //!
+  //! @param[in] init_value
+  //!   Initial value to seed the inclusive scan (and is assigned to `*d_out`)
```
Comment (bernhardmgruber) [inline cub/cub/device/device_scan.cuh:1228]: Question: is this a copy & paste mistake? IIRC, an inclusive scan writes to the first output position `scan_op(initial, d_in[0])`

**Review by bernhardmgruber — COMMENTED** [review]
Here are a few more comments. Otherwise, LGTM. I assume @gevtushenko may have further comments, since he is more intimate with the implementation.

File: cub/cub/device/dispatch/dispatch_scan.cuh (hunk 14 above)
Comment (gonidelis) [reply cub/cub/device/dispatch/dispatch_scan.cuh:243]: Will break backwards compatibility, no?

File: cub/cub/device/device_scan.cuh (hunk 13 above)
Comment (gonidelis) [reply cub/cub/device/device_scan.cuh:1181]: Nice catch!

File: cub/cub/device/dispatch/dispatch_scan.cuh (hunk 14 above)
Comment (bernhardmgruber) [reply cub/cub/device/dispatch/dispatch_scan.cuh:243]: Fair point. However, it's a loud breakage since users would see a compile error (their custom policy cannot initialize a bool) and the fix is really easy. Furthermore, customizing policies strikes me as a feature for a small set of power users. But let's hear @gevtushenko opinion on this.

Comment (jrhemstad) [reply cub/cub/device/dispatch/dispatch_scan.cuh:243]: We struggle enough as it is with upsetting our users with unintentionally breaking them. We should not make any intentional breaking changes within a minor version release without extreme justification, and I don't think this case meets that threshold.

**Review by gevtushenko — APPROVED** [review]

File: cub/cub/device/dispatch/dispatch_scan.cuh (hunk 14 above)
Comment (bernhardmgruber) [reply cub/cub/device/dispatch/dispatch_scan.cuh:243]: Alright, so we have an answer :)
